token-group/interface/src/instruction.rs

```diff
@@ -0,0 +1,495 @@
+//! Instruction types
+
+use {
+    crate::state::SplTokenGroup,
+    borsh::{BorshDeserialize, BorshSerialize},
+    solana_program::{
+        instruction::{AccountMeta, Instruction},
+        program_error::ProgramError,
+        pubkey::Pubkey,
+    },
+    spl_discriminator::{ArrayDiscriminator, SplDiscriminate},
+    spl_pod::optional_keys::OptionalNonZeroPubkey,
+};
+
+/// Get the slice corresponding to the given start and end range
+pub fn get_emit_slice(data: &[u8], start: Option<u64>, end: Option<u64>) -> Option<&[u8]> {
+    let start: usize = start.unwrap_or(0) as usize;
+    let end = end.map(|x| x as usize).unwrap_or(data.len());
+    data.get(start..end)
+}
+
+/// Instruction data for initializing a new `Group`
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize, SplDiscriminate)]
+#[discriminator_hash_input("spl_token_group_interface:initialize_group")]
+pub struct InitializeGroup<G>
+where
+    G: SplTokenGroup,
+{
+    /// Update authority for the group
+    pub update_authority: OptionalNonZeroPubkey,
+    /// The maximum number of group members
+    pub max_size: Option<u64>,
+    /// Additional state
+    pub meta: Option<G>,
+}
+
+/// Instruction data for updating the max size of a `Group`
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize, SplDiscriminate)]
+#[discriminator_hash_input("spl_token_group_interface:update_group_max_size")]
+pub struct UpdateGroupMaxSize {
+    /// New max size for the group
+    pub max_size: Option<u64>,
+}
+
+/// Instruction data for updating the authority of a `Group`
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize, SplDiscriminate)]
+#[discriminator_hash_input("spl_token_group_interface:update_group_authority")]
+pub struct UpdateGroupAuthority {
+    /// New authority for the group, or unset if `None`
+    pub new_authority: OptionalNonZeroPubkey,
+}
+
+/// Instruction data for initializing a new `Member` of a `Group`
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize, SplDiscriminate)]
+#[discriminator_hash_input("spl_token_group_interface:initialize_member")]
+pub struct InitializeMember {
+    /// The pubkey of the `Group`
+    pub group: Pubkey,
+    /// The member number
+    pub member_number: u64,
+}
+
+/// The type of item to emit
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize)]
+pub enum ItemType {
+    /// Emit the `Group`
+    Group,
+    /// Emit the `Member`
+    Member,
+}
```

Comment:
Actually I originally _had_ in here a proposal for a totally separate interface that handled "extra metas" and "emit", since they might be used across many interface implementations.

So I'm in favor

Reply:
Yeah, let's create a separate interface just for emitting

Reply:
ouuhhh but not extra metas though eh?

Reply:
I didn't really understand what you meant by an interface to handle extra metas if I'm being honest :sweat:

Reply:
It would basically be one instruction: "Give me any interface SPL discriminator, and I'll write extra metas to a buffer for you". Although perhaps this may not need to be an _interface_, but just some processor function you have on your program.

The confusion arises when you start to think whether or not transfer hook should technically include it or just expect it, as explained above 😅

Reply:
ah ok, I understand now, thanks. I think we can punt on that one for now to be honest. I imagine that in the future, especially with runtime v2, we won't need to do so much work outside of the target program, and programs write their own logic for giving the required account metas.

I guess that means I'm advocating for a Noah-style interface for declaring the extra accounts!

Reply:
Punting sounds good. And yep, I guess in a way you kind of are!

![shaq](https://media.giphy.com/media/88iYsvbegSUn9bSTF8/giphy.gif)